---
title: "Lab2: Data types and data structures"
output: 
  learnr::tutorial:
    progressive: TRUE
    allow_skip: FALSE
runtime: shiny_prerendered
date: "`r Sys.Date()`"
tutorial:
  id: "b215.f23.lab1.test"
  version: 0.5
---

```{r setup, include=FALSE}
library(learnr)
library(tidyverse)
library(knitr)
library(ggmosaic)
library(praise)
library(utils)
library(datasets)
library(dplyr)
tutorial_options(exercise.eval = TRUE, exercise.reveal_solution = FALSE)
gradethis::gradethis_setup(
  pass.praise = TRUE,
  fail.encourage = TRUE,
  fail.hint = TRUE
)

```

## Overview

Tasks: 

-   Work through this document
-   Complete DataCamp assignments on data types and data structures
-   Practice, practice, practice!

## Learning outcomes: 

1.  Learn about the different types of data: numeric, integer, logical, character, factor, etc.

2. Understand the importance of the `NA` values in R.

3.  Familiarize yourselves with common types of data structures: vectors, matrices, data.frames, lists

4.  Understand the interactions between vectors and matrices

4.  Use cbind and rbind [remove]

## Data types & Data structures- Why??

This sounds boring and intimidating. Why do we need this?
Well, these are fundamental concepts in programming, and it transcends R -- it applies to any programming language.

* **Data types** define the kind of data that a variable can hold and the operations that can be performed on that data.

* **Data structures** are ways to organize and store data in a structured and efficient manner within a computer program. They allow you to group and manage related pieces of data together. Think of data structures as containers that hold data in specific formats to make it easier to work with that data.

## Data types: How many ways can data come

They help the computer interpret and manipulate the data correctly.

In R:

* The five main ones: Numeric, integer, logical, character, factor
* Extras: date, complex, raw

### Numeric

This is the default computational data type in R. If you enter a number without a decimal point, R will automatically convert it into a numeric data type. Numeric values in R can hold any real number, including both integers and fractions, and even negative numbers. For example, 5, 23.5, and -3 are all numeric values.
```{r, echo=T, eval=F}
pi <- 3.14
print(pi)
```
### Integer

This data type in R is used to hold whole numbers. Unlike numeric values, integers cannot have a decimal point. You can create an integer in R by adding an 'L' suffix to a number, like this: 5L. Without the 'L', R would treat the number as numeric, not integer.

```{r, echo=T, eval=F}
class_number <- 215L
```
### Logical (T/F)

Logical data type in R is used to hold Boolean values: TRUE or FALSE. You can use logical operators like <, >, ==, etc. to generate logical values. For example, the operation 5 > 3 will return TRUE because 5 is indeed greater than 3.
```{r, echo=T, eval=F}
reality <- TRUE
reality2 <- T #also works
```
### Character

Character data type in R is used to hold strings of text. You can create a character string by enclosing text in quotes. For example, "Hello, World!" is a character string.

```{r, echo=T, eval=F}
types_of_data <- c("Numeric", "Interger","Logical", "Complex", "Character", "Raw") 
```

### Complex

This data type in R is used to store complex numbers, which have both real and imaginary parts. The imaginary part is denoted by 'i'. For example, 2+3i is a complex number where 2 is the real part and 3 is the imaginary part. (We won't use this in the course)

```{r, echo=T, eval=F}
complex <- 1 +2i #we dont know when you'd use this i guess if you are mathy idk
```

### Raw (for the real nerds)

Raw data type in R is *_not_* commonly used for beginners. It is used to hold *raw bytes*, which can be anything from the contents of a file to an image. It's different from the other data types in that it doesn't represent something human-readable, but instead something that a machine might need to process. It's often used in more advanced topics like interfacing with other languages or dealing with binary data.

```{r, echo=T, eval=F}
hexcode1<- as.raw(0)
hexcode2 <- as.raw(128)
hexcode3 <- as.raw(255)
```

## Why choose a certain data type?

As mentioned, the data type defines the types of operations that can be done with the data. You wouldn't want to (gasp) take the mean of a vector of cahracters such as `c("mom","Australia", "bologne")`.

Actually, try that and see what happens. Use the function `mean` on the vector above: `print(mean(c('mom','Australia','bologne')))`:

```{r sandbox1, exercise=TRUE, eval = FALSE}
print(mean(c("mom","dad", "bologne")))
```

As you can see, R will still try its best at guessing what you're up to. Unable to calculate the mean of such things, it will return a warning and `NA`. When you see NA in R, it means that the data at that point is missing or undefined. 

### Numeric 
Suppose you are studying the growth rate of a bacterial culture. You have the sizes of the bacterial colonies at different time points and you want to calculate the average growth rate.

```{r, echo=T, eval=F}
growth_data <- c(1.2, 2.3, 3.4, 5.6, 8.9) # Sizes of bacterial colonies at different time points
```


```{r setupnum}
growth_data <- c(1.2, 2.3, 3.4, 5.6, 8.9)
avg_growth<-mean(growth_data)
```
Suppose you want to take the average of these values. Paste `avg_growth <- mean(growth_data)` then print `avg_growth`:

```{r examplenum, exercise = TRUE, exercise.setup = "setupnum"}
                    
```



```{r examplenum-solution}
avg_growth <- mean(growth_data) 
print(avg_growth)

```

```{r examplenum-check}
grade_this_code()
```

We've learned many operations in Lab 1 that are appropriate for numerical or integer data types, such as addition, subtraction etc. Here are a few more:


```{r, echo=T, eval=F}
x<-c(5.2,-7.3, 9.8)
y<-2
print(abs(x)) #absolute values
print(median(x)) #median of all values in x
print(x>y)
```

Practice these a bit here:
```{r sandbox2, exercise=TRUE, eval = FALSE}

```
*Important:* If there are `NA` values present, we need to explicitly tell R how to handle them.

```{r, echo=T, eval=F}
x<-c(5.2,-7.3, 9.8, NA)
median(x) #median of all values in x
mean(x)
max(x)
min(x)
```

Try these out here:
```{r sandbox3, exercise=TRUE, eval = FALSE}

```
As you can see, that's not great. Now train again by adding the parameter na.rm=TRUE to the functions `mean`, `median`, `max`, `min`. E.g. `mean(x, na.rm=T)`. This argument tells R to ignore elements in the vector that are `NA`.


```{r setupsandbox4}
x<-c(5.2,-7.3, 9.8, NA)
```

```{r sandbox4, exercise=TRUE, eval = FALSE, exercise.setup="setupsandbox4"}

```

Hopefully you got the same results as you had when the `NA` element was not present!

### Integer

You are counting the number of cells in different samples under a microscope. This is where integers come into play, as you can't have a fraction of a cell.

Suppose you are studying the growth rate of a bacterial culture. You have the sizes of the bacterial colonies at different time points and you want to calculate the average growth rate.

```{r, echo=T, eval=F}
cell_counts <- c(100, 200, 150, 175, 225) # Number of cells in different samples

```


```{r setupint}
cell_counts <- c(100, 200, 150, 175, 225)
total_cells <- sum(cell_counts)
```

Suppose you want to add up your total number of cells. You could define a variable `total_cells` where you put in the result of the function `sum` on the vector `cell_counts`. Do that and then print the `cell_counts` object:

```{r exampleinteger, exercise = TRUE, exercise.setup = "setupint"}
 
                   
```


```{r xampleinteger-solution}
total_cells <- sum(cell_counts) 
print(total_cells)

```


```{r exampleinteger-check}
grade_this_code()
```

Once again, if your data are plagged with NAs (which is usually the case with real data), you will need to be explicit with R.

```{r, echo=T, eval=F}
cell_counts <- c(100, 200, 150, 175, 225) 
cell_counts_withNA<-c(100, 200, 150, 175, 225, NA)
total_cells <- sum(cell_counts)
total_cells_withNA <- sum(cell_counts_withNA)
total_cells_withNA_correct <- sum(cell_counts_withNA, na.rm=T)
```

Try these for yourself:

```{r sandbox5, exercise=TRUE, eval = FALSE}

```



### Character

In bioinformatics, character strings are often used to represent DNA sequences.



```{r, echo=T, eval=F}
dna_sequences <- c("ATGC", "CGTA", "GATC", "TACG") # DNA sequences

```


```{r setupchr}
dna_sequences <- c("ATGC", "CGTA", "GATC", "TACG") # DNA sequences


```
Suppose you want to look at the first sequence from a long list. Paste the following code to define a character variable and then print it.
`first_sequence <- dna_sequences[1] # DNA sequence`:



```{r examplechr, exercise = TRUE, exercise.setup = "setupchr"}
                    
```


```{r examplechr-solution}
first_sequence <- dna_sequences[1] # Get the first DNA sequence
print(first_sequence)

```

```{r examplechr-check}
grade_this_code()
```


### Logical (also called boolean)
These are always defined as `TRUE` or `FALSE`. 

```{r, echo=T, eval=F}
a <- c(1,2,3) #suppose this is your data and you misenter the first number
b <- c(2,2,3) #and you want to make sure your data matches a previous result
which_entries_match <- a==b #paste this into the sandbox and call print() on the vector to look at the result
```


```{r setupbool}
a <- c(1,2,3) 
b <- c(2,2,3)

```


```{r examplebool, exercise = TRUE, exercise.setup = "setupbool"}
                    
```


```{r examplebool-solution}
which_entries_match <- a==b
print(which_entries_match)
```

```{r examplebool-check}
grade_this_code()
```

### Factors

TO DO!!!


## Data Structures

### Vectors
You already learned about vectors in the last lesson. They are a  sequence of data elements of the same basic type. They can be numeric, integer, complex, character, or logical. They are **one-dimensional.**

R loves vectors and tends to treat everything like a vector: an object with only one element is a vector. E.g., `a<-9` is simply a vector of length `1`.

Check this is true by asserting object `a` is a vector using the function `is.vector`. Next, check its length with the function `length()`:

```{r setupvec1}
a<-9

```

```{r examplevec1, exercise = TRUE, exercise.setup = "setupbool"}

                    
```

```{r examplevec1-solution}
is.vector(a) # use function is.vector on object a
length(a) #use funciton length on object a
```

```{r examplevec1-check}
grade_this_code()
```


### Matrices

A matrix in R is a **two-dimensional** data structure, where each element is of the same atomic type. You can think of a matrix as a collection of vectors arranged side by side (for columns) or one on top of another (for rows). Here's an example of a numeric matrix. 

```{r matrix1, exercise = F }
numeric_matrix <- matrix(1:6, nrow = 2, ncol = 3) 
print(numeric_matrix)
```



## Interaction between Vectors and Matrices:

Vectors and matrices interact in various ways in R. Here are a few examples:

-You can create a matrix from vectors. Each vector can become a row or a column of the matrix.

-You can extract a row or a column from a matrix as a vector.

-You can use a vector to index a matrix, meaning to select certain rows or columns from the matrix.

-You can perform arithmetic operations between a matrix and a vector. In this case, R will often apply the operation element-wise, meaning it will operate on each corresponding pair of elements from the matrix and the vector. If the vector is shorter than the matrix, R will recycle the elements of the vector to match the length of the matrix.

### Creating a matrix from vectors:
You can use the cbind() function to combine vectors as columns of a matrix, or rbind() to combine vectors as rows.
```{r matrix2, exercise = F }
# Create two vectors
vector1 <- c(1, 2, 3)
vector2 <- c(4, 5, 6)

# Combine vectors as columns to create a matrix
matrix1 <- cbind(vector1, vector2)

# Combine vectors as rows to create a matrix
matrix2 <- rbind(vector1, vector2)

```
print them both to compare
```{r setupmat1}
vector1 <- c(1, 2, 3)
vector2 <- c(4, 5, 6)
matrix1 <- cbind(vector1, vector2)
matrix2 <- rbind(vector1, vector2)

```


```{r examplemat1, exercise = TRUE, exercise.setup = "setupmat1"}
                    
```



```{r examplemat1-solution}
print(matrix1)
print(matrix2)

```

```{r examplemat1-check}
grade_this_code()
```


### Extracting a row or a column from a matrix as a vector:
You can do this using indexing. In R, you use square brackets [] to index a matrix. The format is [row, column]

```{r mat2, exercise = F}
# Create a matrix
matrix3 <- matrix(1:9, nrow = 3)

# Extract the first row as a vector
row1 <- matrix3[1, ]

# Extract the first column as a vector
col1 <- matrix3[, 1]
```


```{r setupmat2}
matrix3 <- matrix(1:9, nrow = 3)
row1 <- matrix3[1, ]
col1 <- matrix3[, 1]
```

now print the matrix, row1 and col1 and examine

```{r examplemat2, exercise = TRUE, exercise.setup = "setupmat2"}
 


```



```{r examplemat2-solution}
 print(matrix3)    
  print(row1)
  print(col1)
```


```{r examplemat2-check}
grade_this_code()
```



### Performing arithmetic operations between a matrix and a vector:
R will apply the operation element-wise, and will recycle elements of the vector if necessary.


```{r mat3, exercise = F}
# Create a matrix
matrix4 <- matrix(1:6, nrow = 2)

# Create a vector
vector4 <- c(1, 2, 3)

# Add the vector to each row of the matrix
result4 <- matrix4 + vector4
```


```{r setupmat3}
matrix4 <- matrix(1:6, nrow = 2)
vector4 <- c(1, 2, 3)
result4 <- matrix4 + vector4

```

print the result

```{r examplemat3, exercise = TRUE, exercise.setup = "setupmat3"}
                    
```


```{r examplemat3-solution}
print(result4)
```

```{r examplemat3-check}
grade_this_code()
```
        

## Other ways you might see data

### Lists 
A list in R is an object that contains elements of different types - like numbers, strings, vectors, and even other lists. Lists are very flexible and are often used when data can't be held in other types of data structures. 

```{r list1, exercise = F}
# Create a list that holds information about a species
species_info <- list(
  "SpeciesName" = "Panthera leo",
  "CommonName" = "Lion",
  "Family" = "Felidae",
  "GenomeSize" = 2.8,  # in Giga Base Pairs
  "Diet" = c("Meat")
)

```


```{r setuplist1}
# Create a list that holds information about a species
species_info <- list(
  "SpeciesName" = "Panthera leo",
  "CommonName" = "Lion",
  "Family" = "Felidae",
  "GenomeSize" = 2.8,  # in Giga Base Pairs
  "Diet" = c("Meat")
)


```

print the result and use double square brackets to select the third item from the list

```{r list1ex, exercise = TRUE, exercise.setup = "setuplist1"}
                    
```


```{r list1ex-solution}
print(species_info)
species_info[[3]]
```

```{r list1ex-check}
grade_this_code()
```
        
### Arrays
An array in R is a data structure for storing homogeneous data (i.e., data of the same type). It is a multi-dimensional structure that can hold values of the same atomic type (numeric, logical, character, etc). It can be viewed as a multi-dimensional version of matrices. An array in R is a multi-dimensional data structure that holds values of a single type. You can think of an array as a multi-dimensional table of data, with each dimension being a different attribute.


```{r array1, exercise = F}
# Create an array that holds gene expression data for 3 genes across 4 samples
gene_expression <- array(c(5,3,4,9,2,8,7,1,6,2,3,6,9, 7, 3, 1, 4, 3, 3, 7, 0, 2, 9, 4), dim = c(3,4,2))
dimnames(gene_expression) <- list(
  "Gene" = c("Gene1", "Gene2", "Gene3"),
  "Sample" = c("Sample1", "Sample2", "Sample3", "Sample4"),
  "Trials" = c("Trial1", "Trial2")
)


```


```{r setuparray1}
# Create an array that holds gene expression data for 3 genes across 4 samples
gene_expression <- array(c(5,3,4,9,2,8,7,1,6,2,3,6,9, 7, 3, 1, 4, 3, 3, 7, 0, 2, 9, 4), dim = c(3,4,2))
dimnames(gene_expression) <- list(
  "Gene" = c("Gene1", "Gene2", "Gene3"),
  "Sample" = c("Sample1", "Sample2", "Sample3", "Sample4"),
  "Trials" = c("Trial1", "Trial2")
)


```

print the result

```{r array1ex, exercise = TRUE, exercise.setup = "setuparray1"}
                    
```


```{r array1ex-solution}
print(gene_expression)
```

```{r array1ex-check}
grade_this_code()
```

Now, try and extract only the matrix which contains the data for trial 2. If you remember to extract columns and rows of a matrix we used the format col <- matrix[,1] and row <- matrix[1,] respectively. Arrays work the exact same way but because they are comprised of matrices that are glued together along the third dimension, you can call successive matrices as following: matrix1 <- array[,,1], matrix2 <- array[,,2], etc.

Save trial 2 as a new matrix named trail2.

```{r array1ex2, exercise = TRUE, exercise.setup = "setuparray1"}
                    
```


```{r array1ex2-solution}
trail2 <- gene_expression[,,2]
```

```{r array1ex2-check}
grade_this_code()
```

Arrays are useful, but niche data structures. A much more common type of data is a data frame.


## Data Frames 

A data frame in R is a table or a two-dimensional array-like structure in which each column contains values of one variable and each row contains one set of values from each column. The data stored in a data frame can be of numeric, factor, or character type. You will use these a lot, this is just a brief introduction!


```{r df1, exercise = F}
## Create a data frame that holds information about various species
species_info <- data.frame(
  "SpeciesName" = c("Panthera leo", "Gorilla gorilla", "Homo sapiens"),
  "CommonName" = c("Lion", "Gorilla", "Human"),
  "Family" = c("Felidae", "Hominidae", "Hominidae"),
  "GenomeSize" = c(2.8, 3.3, 3.2),  # in Giga Base Pairs
  "Diet" = c("Meat", "Herbivore", "Omnivore")
)



```


```{r setupdfex1}
# Create a data frame that holds information about various species
species_info <- data.frame(
  "SpeciesName" = c("Panthera leo", "Gorilla gorilla", "Homo sapiens"),
  "CommonName" = c("Lion", "Gorilla", "Human"),
  "Family" = c("Felidae", "Hominidae", "Hominidae"),
  "GenomeSize" = c(2.8, 3.3, 3.2),  # in Giga Base Pairs
  "Diet" = c("Meat", "Herbivore", "Omnivore")
)




```

Print the dataframe and on the next row practice using the structure dataframe$columnname to select the column named "Family" and save that as a vector called "family". You can also use square brackets but $ allows you to see the name of the column you are extracting without counting which column you're interested in in the original dataframe.

```{r dfex1, exercise = TRUE, exercise.setup = "setupdfex1"}
                    
```


```{r dfex1-solution}
print(species_info)
family <- species_info$Family
```

```{r dfex1-check}
grade_this_code()
```
   
Dataframes are useful because they can combine multiple different types of data into a single column. This is a dataset showing the hominin (early human) species name and hypothetical genome sizes for these early hominins (which we have no way of knowing yet).

```{r df2, exercise = F}
## Create a data frame that holds information about various species
species_info <- data.frame(
  "Genus" = c("Ardipithecus", "Australopithecus","Australopithecus","Australopithecus","Australopithecus","Australopithecus", "Homo","Homo","Homo","Homo","Homo","Homo","Homo","Paranthropus", "Paranthropus", "Sahelanthropus"),
  "Species" = c("ramidus", "afarensis", "africanus", "anamensis", "garhi", "sediba", "erectus", "floresiensis", "habilis", "heidelbergensis", "naledi", "rudolfensis", "sapiens","boisei", "robustus", "tchadensis"),
  "Genome length" = c(2.80, 2.86, 2.90, 2.79, 3.48, 2.68, 3.29, 3.30, 3.11, 3.30, 2.80, 3.10, 3.25,3.17, 3.29, 3.01)  # in Giga Base Pairs
)



```


```{r setupdfex2}
## Create a data frame that holds information about various species
species_info <- data.frame(
  "Genus" = c("Ardipithecus", "Australopithecus","Australopithecus","Australopithecus","Australopithecus","Australopithecus", "Homo","Homo","Homo","Homo","Homo","Homo","Homo","Paranthropus", "Paranthropus", "Sahelanthropus"),
  "Species" = c("ramidus", "afarensis", "africanus", "anamensis", "garhi", "sediba", "erectus", "floresiensis", "habilis", "heidelbergensis", "naledi", "rudolfensis", "sapiens","boisei", "robustus", "tchadensis"),
  "Genome_length" = c(2.80, 2.86, 2.90, 2.79, 3.48, 2.68, 3.29, 3.30, 3.11, 3.30, 2.80, 3.10, 3.25,3.17, 3.29, 3.01) # in Giga Base Pairs
)

australopithecus_genome_length<-filter(species_info, Genus == "Australopithecus")

```

In the following box, seperate all the rows with "Australopithecus" in the "Genus" column using the function "filter" from the dplyr package. dplyr has a whole host of useful functions that make working with dataframes conveinent. Filter is just the tip of the iceberg. Follow the format: filter(dataframe_name, columnname == "Australopithecus") and save the results as australopithecus_genome_length.

```{r dfex2, exercise = TRUE, exercise.setup = "setupdfex2"}
                    
```


```{r dfex2-solution}
australopithecus_genome_length<-filter(species_info, Genus == "Australopithecus")
```

```{r dfex2-check}
grade_this_code()
```

Now, using the dataframe you made (australopithecus_genome_length), use the function "summary" on the column of that dataframe which contains the genome information to calculate how variable hominin genomes are. Save that result into the object "summary_results" and print out that object.

```{r dfex3, exercise = TRUE, exercise.setup = "setupdfex2"}
                    
```


```{r dfex3-solution}
summary_results <-summary(australopithecus_genome_length$Genome_length)
print(summary_results)
```

```{r dfex3-check}
grade_this_code()
```



## The quiz
This is just graded for completion! First part is basic questions like the tutorial. Feel free to go back to previous sections. You won't lose your progress unless you click "start over".
 
### 1-integer vector
Consider the following integer vector that represents the quantity of different species observed in a field study: species_counts <- c(10, 20, 15, 30, 25). Let's say the researchers are particularly interested in species with observed counts of 20 or below.

Which species are of particular interest? Use logical comparison to create a vector of TRUE/FALSE values indicating which species meet the criteria.

```{r numvecex, exercise = TRUE}
  species_counts <- c(10, 20, 15, 30, 25)
                  
```


```{r numvecex-solution}
species_counts <- c(10, 20, 15, 30, 25)
species_counts < 20
```

```{r numvecex-check}
grade_this_code()
```





### 2- list
You have a list that holds various information about a species,
Is the species endangered? Use list indexing to retrieve the information.


```{r listexam, exercise = TRUE}
species_info_list <- list("SpeciesName" = "Panthera leo", "Population" = 20000, "Endangered" = TRUE)                  
```


```{r listexam-solution}
species_info_list <- list("SpeciesName" = "Panthera leo", "Population" = 20000, "Endangered" = TRUE)
species_info_list$Endangered
```

```{r listexam-check}
grade_this_code()
```




### 3- Cbind/rbind
You have two vectors representing the diameters of two different types of cells observed under a microscope.  Combine these two vectors into a matrix where each column represents a different cell type. name the matrix cell_diameters, then print it. 

```{r bindex1, exercise = TRUE}
cell1_diameters <- c(10, 12, 11, 13, 12)  
cell2_diameters <- c(15, 17, 16, 18, 17)

```


```{r bindex1-solution}
cell1_diameters <- c(10, 12, 11, 13, 12)  
cell2_diameters <- c(15, 17, 16, 18, 17)
cell_diameters<-cbind(cell1_diameters, cell2_diameters)
print(cell_diameters)
```

```{r bindex1-check}
grade_this_code()

```

You have two vectors representing the heights of two different types of plants measured in a field study. Combine these two vectors into a matrix where each row represents a different plant type. name the matrix plant_heights, then print it.


```{r bindex2, exercise = TRUE}
plant1_heights <- c(100, 120, 110, 130, 120) 
plant2_heights <- c(150, 170, 160, 180, 170)

```


```{r bindex2-solution}
plant1_heights <- c(100, 120, 110, 130, 120) 
plant2_heights <- c(150, 170, 160, 180, 170)
plant_heights<-rbind(plant1_heights, plant2_heights)
print(plant_heights)
```

```{r bindex2-check}
grade_this_code()
```


## Quiz Part 2: Working with real data



Now we are going to practice working with an actual data set. We will be using one from the datasets package called chickweights. It's from an experiment on the effect of diet on early growth of chicks. If you want to learn more about the dataset type ??chickweights into your console. 

### Lets look at the data
The datasets package already has the data stored as a data frame. If it wasn't you could call data.frame to turn it into one. Rename the data chicks by assigning the code already loaded to that name. 
It's long, so we don't want to print the whole thing. Use a specific function to call just the first 6 rows.

```{r chicks1, exercise = TRUE}
datasets::ChickWeight

```


```{r chicks1-solution}
chicks<-datasets::ChickWeight
head(chicks)
```

```{r chicks1-check}
grade_this_code()
```

### Analyzing the data
Lets see which diet worked best! This data shows us the same chicken at multiple times. If we only want to see the final outcome, we can filter for the final weight. Using the dplyr filter function to select only chicks with time 21.
name this resulting dataframe FinalChicks. Then print it. 
```{r setupchicks}
chicks<-datasets::ChickWeight
FinalChicks<-filter(chicks, Time == "21")


```



```{r chicks2, exercise = TRUE, exercise.setup = "setupchicks"}

```


```{r chicks2-solution}
FinalChicks<-filter(chicks, Time == "21")
print(FinalChicks)
```

```{r chicks2-check}
grade_this_code()
```


Now lets get the mean for each diet. Use the pattern below calculate the mean weight for the chicks in each of the 4 diets. 

```{r chicks3, exercise = TRUE, exercise.setup = "setupchicks"}
diet1<-filter(FinalChicks, Diet == "1")
mean(diet1$weight)

```


```{r chicks3-solution}
diet1<-filter(FinalChicks, Diet == "1")
mean(diet1$weight)

diet2<-filter(FinalChicks, Diet == "2")
mean(diet2$weight)

diet3<-filter(FinalChicks, Diet == "3")
mean(diet3$weight)

diet4<-filter(FinalChicks, Diet == "4")
mean(diet4$weight)

```

```{r chicks3-check}
grade_this_code()
```



Isn't that a lot of work to get 4 simple means? Next week you are going to learn how to do all that in one go! Here's how you'd do it. Print the dataframe from this code to see how fast it is!

```{r chicks4, exercise = TRUE, exercise.setup = "setupchicks"}
avg_chick_wgts<- chicks %>%
filter(Time == "21") %>% 
group_by(Diet) %>%
summarize(mean_weight = mean(weight, na.rm = TRUE))
```


```{r chicks4-solution}
avg_chick_wgts<- chicks %>%
filter(Time == "21") %>% 
group_by(Diet) %>%
summarize(mean_weight = mean(weight, na.rm = TRUE))

print(avg_chick_wgts)
```

```{r chicks4-check}
grade_this_code()
```
