---
title: "Lesson 7-Key!"
author: "YOUR NAME"
date: "`r Sys.Date()`"
output: pdf_document
header-includes:
  \usepackage{fvextra}
  \DefineVerbatimEnvironment{Highlighting}{Verbatim}{breaklines,commandchars=\\\{\}}
---
```{r setup, echo=F}
knitr::opts_chunk$set(echo=T, eval=T)
options(tinytex.verbose = TRUE)
suppressMessages(library(dplyr))
#load ggplot package
suppressMessages(library(ggplot2))
```


## Learning goals

* Using R's built-in loop functions: `apply`, `lapply` and contrasting them with `for` loops

* Conditional statements

* Investigate sampling error

* Visualize confidence intervals.

* Adding error bars to plots


## Apply and lapply

Last week you learned about loops, which are standard in all programming languages. However, R has some built in functions that perform similar operations and are a bit more simple to use. They belong to the "apply" family. Here are some differences between apply and lapply
Differences between lapply and apply:
Data Types:

* `lapply` works on lists and vectors.
* `apply` works on matrices, arrays, and data.frames.

Return Type:

* `lapply` always returns a list.
* `apply` can return a vector, matrix, or array, depending on the function being applied and the input's dimensions.

Use Case:

* Use `lapply` when you have a list or vector and you want to apply a function to each of its elements.
* Use `apply` when you have a matrix or array or data.frame and you want to apply a function across its rows or columns.


There are many other functions in the apply family but we are just focusing on these two.
 
The apply function works with three main arguments: x, margin, and function. X simply determines which array, data frame, or matrix you will be manipulating, as it does in most functions. Function, also quite obviously, signifies which function you'd like to apply to the data. Margin, the second argument, is the most tricky. Margin tells apply which rows and columns to perform the function on. This can be in the normal indexing form, with 1 being rows in a matrix while 2 is columns and c(1,2) is rows and columns, or if the columns and rows are names, you can use a character vector indicating the names of the columns of interest.

#### 1. Let's start super basic. Say you have a matrix, `mat <- matrix(1:12, nrow=3)`. You want to find the sum of each column. paste the matrix first then use apply to find the sum. Name it something then print the result


```{r ex1}
#paste mat

#use apply to take the sum of the columns 

#print your results

```

##### 2. Use lapply to find how many characters are in each sequencing fragment. 

`lapply` works much the same as `apply` except the `margin` argument is not included, as lists and vectors are one dimensional. Here let's say you have a list of fragments from sequencing a genome you're interested in studying. You want to know how many base pairs long each fragment is. We can use `lapply` and the `nchar` function to determine this.
Here are your fragments, get counting
`fragments_list <- list("AGGCTATGACC", "TCCGAT", "TTTTAGATTCATGACCCAG", "GATCA", "AAGCTAGATCAA", "GCCCCAT", "CCCTAG")`
Save your results as an object and then print.

```{r ex2}
#paste fragments_list

#use lapply and save the results

#print your results

```
What is different about this output? 
Correct, it's a list!


#### 3. Comparing `apply` and `for` loops 

Let's compare using `apply` and `for` loops with an example adapted from real biological research. Recall the phagocyte data from Dr. Williamson's lab which we looked at in Lab Assignment 1. If you need a refresher, here is a link to a description of the project and original data. Below, is a model of that data on a much smaller and simplified scale. Imagine that you are a researcher working in Dr. Williamson's lab and you are tasked with observing a subset of the phagocyte replicates. The matrix below contains your sample with the three replicates and the amount of cells they ate at each level of surface stiffness (pmds).    

```{r setupex3}
phagocyte <- matrix(c(3, 9, 7, 8, 
                      7, 2, 1, 1, 
                      6, 5, 8, 4), 
                 nrow=3, byrow=TRUE)
colnames(phagocyte) <- c("pmds_15", "pmds_69", "pmds_300", "pmds_2280")
rownames(phagocyte) <- c("rep_1", "rep_2", "rep_3")
print(phagocyte)
```


First use the new functions you learned to find the average amount of cells eaten for each student. 



```{r. ex3.1}
#use apply


```

Now use a for loop. Remember to create blank variables first to store your information. You can call this one `averages`. Hint



```{r ex3.2}
#first count the number of cell counts in phagocytes and save it as num_cells

num_cells<-nrow(phagocytes)
#create blank variable called averages. Make it numeric with as many elements as num_cells

averages<-c()

#create your for loop to iteratively calculate the mean cells eaten for each replicate and store it in different positions of the object averages

#name the object averages using the rownames of object scores



```



* Which of these was easier for you? Which replicate had the most cells eaten?


* Does one level of surface stiffness allow more phagocytosis than others? Lets do the process again to find the average (mean) for each stiffness level


First use `apply` and save the results in an object called `cell_averages_apply`:

```{r ex3.3}
#create object

#print your object

```

Then use a `for` loop and save the results in an object called `cell_averages_loop`:

```{r ex3.4}

#first count the number of stiffness levels in the scores matrix and save it as num_stiffness:

#create blank variable called cell_averages_loop. Make it numeric with as many elements as num_stiffness

#create your for loop to iteratively calculate the mean cells eaten per stiffness level




#name the object cell_averages_loop using the colnames of object phagocyte

#print your object
```


And now you understand the joy of the apply family! 

## Conditional Statements
Conditional Statements are a key part of any programming language. 

### 1. if Statement:
The `if` statement evaluates a condition, and if that condition is `TRUE`, it executes the code inside the statement.

```{r }
x <- 10

if (x > 5) {
  print("x is greater than 5")
}


```

In the above code, since `x` is indeed greater than 5, the message "x is greater than 5" will be printed.

The value of x doesn't chnage how the conditional statement works, just its output. Change `x` to a few different numbers to see how the output changes!

### 2. if-else Statement:
There might be cases where you want to execute one set of instructions if the condition is `TRUE` and another set if it's `FALSE`. This is where the `if-else` statement comes in.



```{r}
x <- 3

if (x > 5) {
  print("x is greater than 5")
} else {
  print("x is not greater than 5")
}

```

You can keep adding else statements!

### 3. if-else if-else Ladder:

For multiple conditions, we can use an `if-else` ladder.


```{r}
x <- 5

if (x > 10) {
  print("x is greater than 10")
} else if (x > 5) {
  print("x is greater than 5 but less than or equal to 10")
} else {
  print("x is less than or equal to 5")
}

```

And so on and so on. This is really useful for making code do complicated things!

Change `x` to a few different numbers to see how the output changes!

##### 4. Write a basic if statement that prints "this number is positive" when that is true. 

Try some numbers!


```{r ex4.1}
#type your code here


```



Now add an `else if` statement that tells you if the number is negative, and an else that tells you if its 0.

```{r ex4.2}
#type your code here


```


Amazing! You now understand basic conditionals! Let's combine this with loops. Here is an example of a loop that contains an if statement 
Lets say you have a bunch of numbers and you want to know how many are positive. 


```{r}
# a vector of numbers
numbers <- c(-5, 3, 8, -2, 0, 7, -9)

# a counter variable that will change each time the loop cycles
positive_count <- 0
#the for loop
for (num in numbers) { #take note of the nested brackets!
  if (num > 0) {
    positive_count <- positive_count + 1 #add 1 to positve_count if number if positive. Otherwise do nothing
  }
}

```

This loop adds one to the count for each positive number.

#### 5. Write a loop telling me the sum of all the even numbers.

Insert Hint
```{r ex5}
# I gave you a bunch of random numbers. 
numbers <- c(1, 2, 3, 4, 5, 6, 7, 8, 9, 10)
# start your code here
# create a counter variable called sum_even



```


## More work in sampling
Now we are going to return to our example from last class, the human genes data set. We are going to graph more sampling distributions and add error bars to some plots. The sampling distributions we created last week allow us to find the confidence interval for our means. This section demonstrates using loops in R to find statictical meausres such as these, a more in depth explanation of these terms may be found here (insert link)


#### 6. Revisiting human_genes

Last class we compared the means of differently sized samples. Generally, the larger your sample, the closer it will be to the true population mean. 

The human gene data set is a bit weird though. Because it has a few genes that are huge, a large sample size might be more likely to include one of those giant genes and mess up your mean. 

For a real experiment you would just deal with that, we touched on some of these options in the previous lab. For our purposes, you should filter the data set to only include genes smaller than 15000 base pairs. Load in your data and do this. You can keep it's name the same.
.

```{r ex6.1}
#load the readr library

#read in human_genes.csv into human_genes using read_csv.


#filter dataset to only contain genes of size <=15000 and save it as human_genes2



```


Run the loop you wrote last week and investigate the means now that we have filtered the data. This was the loop where you took 100 samples of three different sizes: 10, 100, 1000, with replacement, and calculated their means. Run that loop again here with this filtered dataset but take 1000 samples instead of 100. Look at sample sizes of 10, 50, and 500. What do you see?


```{r ex6.2}
#calculate the mean gene length for human_genes


#create empty numeric vectors with 1000 positions for the means. Call them mean_values_10, mean_values_50, mean_values_500



#start your loop here. Take samples with replacement, sizes 10, 50, 500 and take the median of those samples


#take the mean of means (the grand mean) of these three vectors and save them to a new vector called grand_means


#compare these three means with the real mean of the filtered dataset


```

#### 7. What's up with these samples?


1. Start by finding the standard error of the mean, which is your the standard deviation of the sampling distribution. Your three sampling distributions were created above and are called mean_values_10, mean_values_50, and mean_values 500.

```{r ex7.1}
#call them sem_10, sem_50, sem_500


```

Congratulations! You just found the standard error by taking the standard deviation of the sampling distributions! 

Now, think about which one of them should have the lowest SEM? Now check if that's true.  Does it match your expectations?


2. Now lets find the 95% confidence intervals. To accomplish this, we will use the `quantile` function. The quantile function, as the name suggests, provides the numbers in a distribution which represent the bounds of a certain percentage of the data. For example, if you find the number which corresponds to a quantile of 0.3, all the numbers above that value in the distribution represent 70% of the data. Here, we will use it to find the middle 95% of the data, aka the 95% CI. `quantile` takes the arguments `x` and `probs = seq()`. `x`, as per usual, dictates the data the function will be performed on, in this case a sampling distribution. `probs = seq ()` takes the quantiles you want to find numbers for, if you put two in you can find upper and lower bounds for some middle percentage. (hint) 

```{r ex7.2}
#call them CI_10, CI_500, CI_500


```


Now we have these confidence intervals, and you didn't need a crazy formula or assumptions about the mean being normally distributed! You simply took samples with replacement and calculated your confidence intervals empirically - aka, you used bootstrapping! 




#### 8. Lets add the CIs to a graph!

Lets first make a data frame of all the information we will need for the chart. You need the following columns: SampleSize, MeanValues,  UpperCI and LowerCI. You can use the ones you've calculated above.


```{r ex8.1}

#create a data.frame called df containing the columns listed above.




```


Amazing! Now make a strip chart using this data frame. Don't worry about CI yet. Just make a simple plot using geom_point() for each of the means, grouped by sample size. Add labels though. See the expected output document for inspiration.


```{r ex8.2}
#first, make sure SampleSize is treated as a factor


#now create your plot code here and save it as p1


#print p1, aka, your plot

```

Gorgeous. Now add your error bars. Use `geom_errorbar` to see the difference in those sample sizes. Add geom_errorbar after geom_point(). Inside aes(), use ymin and ymax with the lower and upper CIs, respectively.

```{r ex8.3}
#create your plot code here and save it as p2


#print p2, aka, your plot


```

Those are super different! As you can see, larger sample sizes create narrower confidence intervals. This is **CRUCIAL** for your own data analysis. 


#### 9. Another graph
Above we plotted only the grand means of the sampling distributions. But what do these sample distributions look like compared to a normal distribution? R has a built in plot (gasp it's not ggplot2) called a density plot that will show us. 
First, use the density function to find the density of each of your sample means. Name them something you'll remember.
PS: ggplot also has an option for density plots. 


```{r 9.1}
#type your code here calling the function density on each of your mean_10, etc vectors and call them dense10, etcc\


```

Now use the R function "plot" to plot each of these. 

```{r}



```

You see as the n increases they get more normal!


### The end!
